src/ParseNode.js

```diff
@@ -61,309 +31,535 @@ export type SymbolParseNode =
     ParseNode<"textord">;
 
 // Union of all possible `ParseNode<>` types.
-// Unable to derive this directly from `ParseNodeTypes` due to
-// https://github.com/facebook/flow/issues/6369.
-// Cannot use `ParseNode<NodeType>` since `ParseNode` is not strictly co-variant
-// w.r.t. its type parameter due to the way the value type is computed.
-export type AnyParseNode =
-    SymbolParseNode |
-    ParseNode<"array"> |
-    ParseNode<"color"> |
-    ParseNode<"color-token"> |
-    ParseNode<"op"> |
-    ParseNode<"ordgroup"> |
-    ParseNode<"size"> |
-    ParseNode<"styling"> |
-    ParseNode<"supsub"> |
-    ParseNode<"tag"> |
-    ParseNode<"text"> |
-    ParseNode<"url"> |
-    ParseNode<"verb"> |
-    ParseNode<"accent"> |
-    ParseNode<"accentUnder"> |
-    ParseNode<"cr"> |
-    ParseNode<"delimsizing"> |
-    ParseNode<"enclose"> |
-    ParseNode<"environment"> |
-    ParseNode<"font"> |
-    ParseNode<"genfrac"> |
-    ParseNode<"horizBrace"> |
-    ParseNode<"href"> |
-    ParseNode<"htmlmathml"> |
-    ParseNode<"infix"> |
-    ParseNode<"kern"> |
-    ParseNode<"lap"> |
-    ParseNode<"leftright"> |
-    ParseNode<"leftright-right"> |
-    ParseNode<"mathchoice"> |
-    ParseNode<"middle"> |
-    ParseNode<"mclass"> |
-    ParseNode<"mod"> |
-    ParseNode<"operatorname"> |
-    ParseNode<"overline"> |
-    ParseNode<"phantom"> |
-    ParseNode<"hphantom"> |
-    ParseNode<"vphantom"> |
-    ParseNode<"raisebox"> |
-    ParseNode<"rule"> |
-    ParseNode<"sizing"> |
-    ParseNode<"smash"> |
-    ParseNode<"sqrt"> |
-    ParseNode<"underline"> |
-    ParseNode<"xArrow">;
+export type AnyParseNode = $Values<ParseNodeTypes>;
```

Comment:
Is it possible to deduplicate `mode` and `loc` using intersection types(`&`)?

Reply:
```diff
diff --git a/src/ParseNode.js b/src/ParseNode.js
index 26b2bbe..7f8cc10 100644
--- a/src/ParseNode.js
+++ b/src/ParseNode.js
@@ -7,7 +7,14 @@ import type {Token} from "./Token";
 import type {Measurement} from "./units";

 export type NodeType = $Keys<ParseNodeTypes>;
-export type ParseNode<TYPE: NodeType> = $ElementType<ParseNodeTypes, TYPE>;
+type ParseNodeBase ={|
+    mode: Mode,
+    loc?: ?SourceLocation,
+|}
+export type ParseNode<TYPE: NodeType> = {|
+    ...$ElementType<ParseNodeTypes, TYPE>,
+    ...ParseNodeBase
+|};

 export type LeftRightDelimType = {|
     type: "leftright",
@@ -31,20 +38,19 @@ export type SymbolParseNode =
     ParseNode<"textord">;

 // Union of all possible `ParseNode<>` types.
-export type AnyParseNode = $Values<ParseNodeTypes>;
+export type AnyParseNode = {|
+    ...$Values<ParseNodeTypes>,
+    ...ParseNodeBase
+|};
```
seems to work, not sure it's correct.

Reply:
Nice find on the type spread operator!

Regarding the `ParseNode` definition above.
1. I would really like `ParseNode` to be covariant with its `TYPE` parameter. This means not looking up the value type within its definition. I think that might have been the cause of a lot of "___ is not compatible with ___" style errors which are completely unhelpful. Flow seems to be able to deal with an explicit union of explicitly-defined structs quite easily. An alternative is to replicate `...ParseNodeBase` in every definition, which I can do.
2. As a minor nit: if we _not_ make `ParseNodeBase` a _strict_ object type, then every `ParseNode<TYPE>: ParseNodeBase` as well. Not sure if we would every use that fact, though.
3. The `AnyParseNode` as defined above is not technically a union of all the structs and the behavior of Flow's type checking (especially when it gives an error) might get very confusing in such cases. E.g. Flow will allow you to do the following with a union that type it won't with the type defined above.
   ```javascript
   type AllTypes = {| type: "foo", foo: number |} |
                   {| type: "bar", bar: string |};
   function(val: AllTypes)  {
     let len;
     if (val.type === "foo") {
       len = val.foo + 1;
     } else {
       len = val.bar.length;
     }
   }
   ```
   [Playground](https://flow.org/try/#0C4TwDgpgBAggNnAKuCBnKBeKBvAPlUSALigCIAzAe0tIBoorKSA7AVwFsAjCAJylwC+-HPkIQSpTgEMedKNJ4lUwHgEtmAc34CA3ACg95VswDGwVZWZQAjqwAeACgBuUuCXhIUqAJRQcevzgIYCgg5n0-VXIoZ1cAOjFMDCwKalJfbAC-UIgrLBc4OMYoAGooAEYIqCEIOFRoTOycvKgCuIU4sI1gAAsqgT0BoA)

   I was thus hoping to eventually change uses of `checkNodeType` to be simple `type` checks.

Reply:
@ylemkimon Actually, using
```
type ParseNodeBase = {
    mode: Mode,
    loc?: ?SourceLocation,
};

type ParseNodeTypes = {
    "array": {|
        type: "array",
        ...ParseNodeBase,
        value: ArrayEnvNodeData,
    |},
    ... // and same on the rest
````
seems to freeze the flow type checker very consistently (it's stuck at 24% progress at 100% CPU utilization). So let's skip that for now...